### ISSUE TITLE
Cannot properly install Pyrosm

### ISSUE THREAD

**Issue by DanielePassabi** [HTenkanen/pyrosm#153]
Hi, I can no longer install and import the library correctly.

**FIRST ATTEMPT**
- First I tried on my main computer, running Windows 10 version 10.0.19043 Build 19043.
- I created a virtual environment using anaconda, installed `geopandas` using `conda install geopandas` and subsequently `pyrosm` using `conda install -c conda-forge pyrosm`, as recommended in the [documentation](https://pyrosm.readthedocs.io/en/latest/installation.html). 
- The installation is successful, but when I try to import the library I get this error:
![image](https://user-images.githubusercontent.com/38936531/141334464-bf0d60b8-5fe1-4810-866e-8b752719a279.png)
- Reinstalling anaconda did not lead to any solution.

**SECOND ATTEMPT**
- I then tried to recreate the problem on my laptop (still windows 10 version 10.0.19043 Build 19043).
- The same error occurs again

**THIRD ATTEMPT**
- So I decided not to use an Anaconda env, but a simple Python virtual env. 
- Still, same error
- Even installing only via pip the problem remains

**FOURTH ATTEMPT**
- I created a virtual machine with Ubuntu 64bit 20.04 
- Installing everything with `conda` (and `conda-forge`) leads to the same problem
- Whether using an anaconda environment or a Python env the problem remains

- Using only pip to install the packages, the installation and import is (apparently) successful, but when you try to run a simple piece of code, the kernel dies (see image below).
![image](https://user-images.githubusercontent.com/38936531/141337617-24daa78a-fc89-4ee3-9308-41e8711e509a.png)
![image](https://user-images.githubusercontent.com/38936531/141337564-5a458fb1-1c27-4494-ba1a-5fc0ace33879.png)

**FIFTH ATTEMPT**
- I finally tried to use Google Colab
- The packages correctly install via pip
![image](https://user-images.githubusercontent.com/38936531/141335724-209ec5f2-1845-47c5-9d46-a37cd919051f.png)
![image](https://user-images.githubusercontent.com/38936531/141335745-da962e84-f9ad-4b52-a600-b1d40cb5b1db.png)
- and it is possible to import them, but with the same warning obtained in Linux (as expected)
```
/usr/local/lib/python3.7/dist-packages/geopandas/_compat.py:115: UserWarning: The Shapely GEOS version (3.9.1-CAPI-1.14.2) is incompatible with the GEOS version PyGEOS was compiled with (3.10.0-CAPI-1.16.0). Conversions between both will be slow.
  shapely_geos_version, geos_capi_version_string
```
- Still, if you try to run the code the kernel dies.
![image](https://user-images.githubusercontent.com/38936531/141335933-956457de-91b2-4c64-8904-7352e6a7d564.png)

PS: on the laptop I have a working environment with both Geopandas and Pyrosm, where the code runs correctly. The problem is that I can no longer replicate such an environment. Even creating a `.yml` file and reinstalling exactly the same packages generates the error. I have tried on multiple devices and OS, but with no success.

I have exhausted every option...

Thank you for your time,
Daniele

**Comment by HTenkanen** [comment]
Thanks @DanielePassabi for reporting this! There has been some updates on the cykhas library lately which is probably the reason for this. Could you try and install an older version of cykhas library before installing pyrosm, and see if that solves the issue?

**Comment by DanielePassabi** [comment]
Thank you so much for your quick answer! @HTenkanen 

I installed `cykhash-1.0.2` beforehand and everything works fine now :)

I don't know if it is possible but it would be a good idea to have this version of `cykhash-1.0.2` downloaded when using `conda-forge`.

**Comment by HTenkanen** [comment]
Yes, I need to update the code base to reflect the latest changes in cykhash for the next release. But yes, freezing the cykhash version to older one is a quick fix for the installation issues.

**Comment by HTenkanen** [comment]
Note to self: 
- There has been breaking changes in Cykhash v2.0.0: use of [different names/signature for maps](https://github.com/realead/cykhash#release-200-09112021) which is the likely reason for this behavior observed in here. 

Todo:
  - Update the logic how the `isin()` lookups are done in `data_filter.pyx` using khash

**Comment by realead** [comment]
@HTenkanen 

The issue is, that the used pyrosm-version was built with cykhash v1.0.2 but runs with cykhash v2.0.0.  One has either to run with cykhash v1.0.2 or to rebuild pyrosm against cykhash2.0.0.

This explains the observations: 

   * pyrosm from conda was built with cykhash v1.0.2 and thus would fail at runtime when installed along with cykhash v2.0.0
   * pyrosm installed from pip would build pyrosm against cykhash v2.0.0 and pyrosm would "work", i.e. the issue seems to be unrelated to cykhash as far as I can tell.

The offending change was the changing of the signature (I got the signature wrong in the first version):

```
cpdef Int64Set_from_buffer(int64_t[:] buf, double size_hint=*)
```
to 

```
cpdef Int64Set Int64Set_from_buffer(int64_t[:] buf, double size_hint=*)
```

It is in so far not breaking, that the old code would **build** and work correctly with the new version. But "breaking" in the sense that code built against old version would not **run** with the new version. I was not aware of this problem until now. 

It is probably best to pin the exact C-extension version used for a build at least as far as cdef-functionality is used.

**Comment by HTenkanen** [comment]
@realead Ah right, thanks for a quick reply and clarifying the change! 👍🏻 (did not know whether there were some larger changes involved). So this is a rather easy fix then, will update the setup accordingly, thanks! 🙂

**Comment by kevinle-1** [comment]
Just leaving a comment containing the error for search engines to crawl as I had spent some time looking for this issue 😅 - Just hopefully to save someone some time down the line. 

```
TypeError: C function cykhash.khashsets.Int64Set_from_buffer has wrong signature (expected PyObject *(__Pyx_memviewslice, int __pyx_skip_dispatch, struct __pyx_opt_args_7cykhash_9khashsets_Int64Set_from_buffer *__pyx_optional_args), got struct __pyx_obj_7cykhash_9khashsets_Int64Set *(__Pyx_memviewslice, int __pyx_skip_dispatch, struct __pyx_opt_args_7cykhash_9khashsets_Int64Set_from_buffer *__pyx_optional_args))
```

**Comment by HTenkanen** [comment]
Hi @DanielePassabi, thanks for reporting this (and to @realead for the quick clarification at the time) — sorry it stayed open for so long.

The root cause was the cykhash 1.0.2 → 2.0.0 transition: the published package was built against cykhash 1.0.2 but ended up running against 2.0.0, which broke the khash lookups. That has since been fixed — `pyrosm` was updated to the cykhash 2.0 API (#155) and now both builds against and requires `cykhash>=2`, so the version mismatch can no longer happen on a current install.

Closing as resolved — thanks again\!
